alluxio-python-library/alluxio/worker_ring.py

```diff
@@ -0,0 +1,193 @@
+                    WorkerNetAddress.DEFAULT_DOMAIN_SOCKET_PATH,
+                ),
+            )
+        except json.JSONDecodeError as e:
+            raise type(e)(
+                f"Provided worker_info is not a valid JSON string {e}"
+            ) from e
+        except AttributeError as e:
+            raise type(e)(
+                f"Provided worker_info must be a bytes-like object {e}"
+            ) from e
+        except Exception as e:
+            raise type(e)(
+                f"Failed to process given worker_info {worker_info} {e}"
+            ) from e
+
+    def dump_main_info(self):
+        return (
+            "WorkerNetAddress{{host={}, containerHost={}, rpcPort={}, dataPort={}, webPort={}, domainSocketPath={}}}"
+        ).format(
+            self.host,
+            self.container_host,
+            self.rpc_port,
+            self.data_port,
+            self.web_port,
+            self.domain_socket_path,
+        )
+
+
+class EtcdClient:
+    PREFIX = "/DHT/DefaultAlluxioCluster/AUTHORIZED/"
+
+    def __init__(self, host="localhost", port=2379):
+        self.host = host
+        self.port = port
+
+    def get_worker_addresses(self):
+        """
+        Retrieve worker addresses from etcd using the specified prefix.
+
+        Returns:
+            list: A list of WorkerNetAddress objects.
+        """
+        # Note that EtcdClient should not be passed through python multiprocessing
+        etcd = etcd3.client(host=self.host, port=self.port)
+        try:
+            return [
+                WorkerNetAddress.from_worker_info(worker_info)
+                for worker_info, _ in etcd.get_prefix(self.PREFIX)
+            ]
+        except Exception as e:
+            raise type(e)(
+                f"Failed to achieve worker info list from ETCD server {self.host}:{self.port} {e}"
+            ) from e
+
+
+class ConsistentHashProvider:
+    def __init__(
+        self,
+        etcd_host,
+        logger=None,
+        num_virtual_nodes=2000,
+        max_attempts=100,
+    ):
+        self.etcd_host = etcd_host
+        self.num_virtual_nodes = num_virtual_nodes
+        self.max_attempts = max_attempts
+        self.logger = logger or logging.getLogger("ConsistentHashProvider")
+        self.is_ring_initialized = False
+
+    def init_worker_ring(self):
+        """Initialize the worker ring using worker addresses from Etcd."""
+        worker_addresses = EtcdClient(self.etcd_host).get_worker_addresses()
+        hash_ring = SortedDict()
+        weight = math.ceil(self.num_virtual_nodes / len(worker_addresses))
+        for worker_address in worker_addresses:
+            worker_string = worker_address.dump_main_info()
+            for i in range(weight):
+                hash_key = self._hash(worker_string, i)
+                hash_ring[hash_key] = worker_address
+        self.hash_ring = hash_ring
+        self.is_ring_initialized = True
+
+    def get_multiple_workers(
```

Comment:
In the current design, would each REST API call trigger a call to the etcd? Could we avoid redundant calls?

Reply:
no, we would only call ETCD and form the hash ring when initializing the AlluxioFileSystem.
There will be only one call and we will not refresh the worker hash ring in this version
```
 def __init__(self, etcd_host, dora_root, options, logger, concurrency=64):
        self.dora_root = dora_root
        self.logger = logger or logging.getLogger("AlluxioRest")
        self.session = self._create_session(concurrency)
        page_size = None
        if self.ALLUXIO_PAGE_SIZE_KEY in options:
            page_size = options[self.ALLUXIO_PAGE_SIZE_KEY]
            self.logger.debug(f"Page size is set to {page_size}")
        else:
            page_size = "1MB"
        self.page_size = humanfriendly.parse_size(page_size, binary=True)
        self.hash_provider = ConsistentHashProvider(etcd_host, self.logger)
        self.hash_provider.init_worker_ring()

```